### ISSUE TITLE
ttnn.sum/mean/max/min/std/var: the registered reduction goldens silently drop the `scalar=` kwarg, so the reference is off by up to 300% at PCC 1.0

### ISSUE THREAD

**Issue by p32929** [tenstorrent/tt-metal#57115]
### Component / Area

`ttnn` reduction goldens — `ttnn/ttnn/operations/reduction.py` (`_create_golden_function`), registered onto `ttnn.sum` / `ttnn.mean` / `ttnn.max` / `ttnn.min` / `ttnn.std` / `ttnn.var`.

### Issue Type (optional)

Bad Outputs

### Observed

The generic-reduction golden factory drops the `scalar=` kwarg. It is absorbed by `**_` and never reaches the torch function, so the golden computes the **unscaled** reduction while the device computes the scaled one.

```python
# ttnn/ttnn/operations/reduction.py @ a4e9bec
def _create_golden_function(torch_function_name):
    def golden_function(input_tensor, dim=None, keepdim=False, **_):   # <-- scalar lands in **_
        torch_function = getattr(torch, torch_function_name)
        ...
        return torch_function(input_tensor, dim=dim, keepdim=keepdim)  # <-- and is never used
```

`scalar` is a live, documented, non-deprecated parameter on all six ops:

```cpp
// ttnn/cpp/ttnn/operations/reduction/generic/generic_reductions_nanobind.hpp:56
scalar (float, optional): A scaling factor to be applied to the input tensor. Defaults to `1.0`.
// :179  nb::arg("scalar") = 1.0f        <-- sum / mean / max / min
// ttnn/cpp/ttnn/operations/reduction/generic/std_var_reductions_nanobind.cpp:82,97
//       nb::arg("scalar") = 1.0f        <-- std / var
```

(For contrast, `correction` on the four generic ops *is* explicitly deprecated on line 57 and has no effect — `scalar` carries no such notice, and `generic_reductions.cpp` threads it all the way into the reduce, including the sign-flip special case on lines 736-745: *"Scaling is applied after reduction, so flip the op for negative scalars"*.)

`ttnn/ttnn/decorators.py:1763` and `:1375`/`:1399` forward the caller's `**kwargs` to the golden verbatim, so `scalar=` does reach `golden_function` — it is simply discarded there.

The repo's own unit tests establish the intended contract and hand-roll the reference rather than calling the registered golden, which is why nothing catches this:

| test | asserted device contract |
|---|---|
| `tests/ttnn/unit_tests/operations/reduce/test_max.py:164` | `ttnn.max(x, dim, scalar=s)` == `torch.amax(s * x, dim)` |
| `tests/ttnn/unit_tests/operations/reduce/test_reduction.py:132` | `ttnn.var(x, dim, scalar=s)` == `torch.var(x * s, dim)` |
| `tests/ttnn/unit_tests/operations/reduce/test_reduction.py:180` | same, `dim=-1`, `Wt>1` |
| `tests/ttnn/unit_tests/operations/reduce/test_reduction_mean.py:99` | `ttnn.mean(x, dim, scalar=s)` |
| `tests/ttnn/nightly/unit_tests/operations/reduction/test_sum_int.py:106` | `ttnn.sum(x, dim, scalar=s)` |

**This also survives the in-flight rewrite.** Open PR #56748 replaces this factory wholesale — it adds a `_ReductionGoldenSpec` table, fixes `correction` for std/var, fixes the `torch.max/min(dim=...)` namedtuple, and fixes `argmax` under `dim=None`. It does not mention `scalar` anywhere; the new signature is `(input_tensor, dim=None, keepdim=False, correction=None, **_)`, so `scalar` is still swallowed. The numbers below are identical for both variants.

### Expected

The golden should apply `scalar` the way the kernel does — to the input, before the reduce:

```python
if scalar != 1.0:
    input_tensor = input_tensor * scalar
```

which reproduces every one of the five test contracts above, including the negative-scalar max↔min flip for free (`amax(-2*x) == -2*min(x)`).

### Magnitude

Pure-Python, no hardware, no ttnn build. `(1,1,32,32)` float32, `torch.manual_seed(0)`, `dim=-1`:

```
torch 2.8.0   input (1,1,32,32) float32, torch.manual_seed(0), dim=-1

--- golden variant A: main @ a4e9bec
op         scalar      max rel err            PCC   note
--------------------------------------------------------
sum          2.00         50.0000%     1.00000000   
sum          0.50        100.0000%     1.00000000   
sum         -2.00        150.0000%    -1.00000000   
mean         2.00         50.0000%     1.00000000   
mean         0.50        100.0000%     1.00000000   
mean        -2.00        150.0000%    -1.00000000   
max          2.00         50.0000%     1.00000000   
max          0.50        100.0000%     1.00000000   
max         -2.00         66.3472%     0.34232361   negative scalar flips the extremum: golden returns a DIFFERENT element
min          2.00         50.0000%     1.00000000   
min          0.50        100.0000%     1.00000000   
min         -2.00         77.9616%     0.34232361   negative scalar flips the extremum: golden returns a DIFFERENT element
var          2.00         75.0000%     1.00000000   expected 4x (scalar^2)
var          0.50        300.0000%     1.00000000   expected 0.25x (scalar^2)
var         -2.00         75.0000%     1.00000000   expected 4x (scalar^2)
std          2.00         50.0000%     1.00000000   expected 2x (|scalar|)
std          0.50        100.0000%     1.00000000   expected 0.5x (|scalar|)
std         -2.00         50.0000%     1.00000000   expected 2x (|scalar|)

--- golden variant B: as rewritten by open PR #56748
op         scalar      max rel err            PCC   note
--------------------------------------------------------
sum          2.00         50.0000%     1.00000000   
sum          0.50        100.0000%     1.00000000   
sum         -2.00        150.0000%    -1.00000000   
mean         2.00         50.0000%     1.00000000   
mean         0.50        100.0000%     1.00000000   
mean        -2.00        150.0000%    -1.00000000   
max          2.00         50.0000%     1.00000000   
max          0.50        100.0000%     1.00000000   
max         -2.00         66.3472%     0.34232361   negative scalar flips the extremum: golden returns a DIFFERENT element
min          2.00         50.0000%     1.00000000   
min          0.50        100.0000%     1.00000000   
min         -2.00         77.9616%     0.34232361   negative scalar flips the extremum: golden returns a DIFFERENT element
var          2.00         75.0000%     1.00000000   expected 4x (scalar^2)
var          0.50        300.0000%     1.00000000   expected 0.25x (scalar^2)
var         -2.00         75.0000%     1.00000000   expected 4x (scalar^2)
std          2.00         50.0000%     1.00000000   expected 2x (|scalar|)
std          0.50        100.0000%     1.00000000   expected 0.5x (|scalar|)
std         -2.00         50.0000%     1.00000000   expected 2x (|scalar|)

Why this has never failed a test: PCC is invariant under a positive scale factor,
so every PCC-gated comparison still reports a perfect score.
  sum   scalar=2.0 -> max rel err   50.00%   PCC 1.0000000000
  var   scalar=2.0 -> max rel err   75.00%   PCC 1.0000000000

self-check: OK

```

Every case is a real divergence, and **every positive-scalar case has PCC exactly 1.0** — PCC is invariant under a positive scale factor, so a 50% or 300% wrong reference sails through any PCC-gated comparison. That is why this has gone unnoticed.

The negative-scalar `max`/`min` rows are worse than a scale factor: the device returns `-2 × min(x)` while the golden returns `max(x)`, i.e. a **different element of the tensor entirely** (PCC 0.342).

### Reproduction Steps

**1. Steps (exact commands)**

```bash
pip install torch          # only dependency; 2.8.0 used above
python repro_reduction_golden_drops_scalar.py
```

The script embeds both golden variants verbatim (today's `main`, and PR #56748's rewrite) and ends in an `assert`-based self-check that also verifies `scalar=1.0` agrees to <1e-3 — so the harness itself is proven sound before any divergence is claimed.

<details>
<summary><code>repro_reduction_golden_drops_scalar.py</code></summary>

```python
#!/usr/bin/env python3
"""
Repro: every registered ttnn generic-reduction golden silently drops the
`scalar=` kwarg, so the golden reference for ttnn.sum / mean / max / min / std /
var is wrong by a factor of `scalar` whenever the caller passes one.

No Tenstorrent hardware and no ttnn build required: the goldens under test are
pure Python and are reproduced here VERBATIM.

  Variant A -- ttnn/ttnn/operations/reduction.py @ a4e9bec (main today)
  Variant B -- the same file as rewritten by open PR #56748, which fixes
               `correction` and the max/min namedtuple but still routes
               `scalar` into `**_`

Device contract, taken from tt-metal's own source and its own unit tests:
  ttnn/cpp/.../generic/generic_reductions.cpp        threads `scalar` into the reduce
  ...generic_reductions_nanobind.hpp:56              "scalar (float, optional): A scaling
                                                      factor to be applied to the input tensor."
  tests/.../reduce/test_max.py:164                   ttnn.max(x, dim, scalar=s)  == torch.amax(s * x, dim)
  tests/.../reduce/test_reduction.py:132             ttnn.var(x, dim, scalar=s)  == torch.var(x * s, dim)
  tests/.../reduce/test_reduction_mean.py:99         ttnn.mean(x, dim, scalar=s)
  tests/.../nightly/.../test_sum_int.py:106          ttnn.sum(x, dim, scalar=s)

Run:  python repro_reduction_golden_drops_scalar.py
"""
from typing import NamedTuple, Optional, Tuple, Union

import torch

torch.manual_seed(0)


# ------------------------------------------------------- Variant A: main @ a4e9bec
def _create_golden_function_main(torch_function_name):
    def golden_function(input_tensor, dim=None, keepdim=False, **_):
        torch_function = getattr(torch, torch_function_name)
        if dim is None:
            if keepdim:
                all_dims = tuple(range(len(input_tensor.shape)))
                return torch_function(input_tensor, dim=all_dims, keepdim=keepdim)
            else:
                return torch_function(input_tensor)
        else:
            return torch_function(input_tensor, dim=dim, keepdim=keepdim)

    return golden_function


# ------------------------------------------------------- Variant B: as per PR #56748
class _ReductionGoldenSpec(NamedTuple):
    torch_name: str
    multi_axis_torch_name: Optional[str] = None
    values_only: bool = False
    passes_correction: bool = False
    widen_unsigned: bool = False
    dim_none_means_flatten: bool = False


_REDUCTION_GOLDEN_SPECS = {
    "mean": _ReductionGoldenSpec("mean"),
    "sum": _ReductionGoldenSpec("sum"),
    "max": _ReductionGoldenSpec("max", multi_axis_torch_name="amax", values_only=True),
    "min": _ReductionGoldenSpec("min", multi_axis_torch_name="amin", values_only=True),
    "var": _ReductionGoldenSpec("var", passes_correction=True),
    "std": _ReductionGoldenSpec("std", passes_correction=True),
}


def _create_golden_function_pr56748(torch_function_name):
    spec = _REDUCTION_GOLDEN_SPECS[torch_function_name]

    def golden_function(input_tensor, dim=None, keepdim=False, correction=None, **_):
        function_kwargs = {"keepdim": keepdim}
        if spec.passes_correction and correction is not None:
            function_kwargs["correction"] = correction
        if dim is None:
            if not keepdim:
                function_kwargs.pop("keepdim")
                return getattr(torch, spec.torch_name)(input_tensor, **function_kwargs)
            dim = tuple(range(len(input_tensor.shape)))
        if spec.multi_axis_torch_name is not None and isinstance(dim, (tuple, list)):
            return getattr(torch, spec.multi_axis_torch_name)(input_tensor, dim=dim, **function_kwargs)
        output = getattr(torch, spec.torch_name)(input_tensor, dim=dim, **function_kwargs)
        if spec.values_only:
            output = output.values
        return output

    return golden_function


# ------------------------------------------------------- device contract
def device(op, x, dim, scalar):
    """What the kernel computes: the reduce is applied to `scalar * x`."""
    return {
        "sum": lambda: scalar * torch.sum(x, dim=dim),
        "mean": lambda: scalar * torch.mean(x, dim=dim),
        "max": lambda: torch.amax(scalar * x, dim=dim),
        "min": lambda: torch.amin(scalar * x, dim=dim),
        "var": lambda: torch.var(x * scalar, dim=dim),
        "std": lambda: torch.std(x * scalar, dim=dim),
    }[op]()


def _t(v):
    return v if isinstance(v, torch.Tensor) else v.values


def pcc(a, b):
    a, b = _t(a).flatten().double(), _t(b).flatten().double()
    a, b = a - a.mean(), b - b.mean()
    d = a.norm() * b.norm()
    return float("nan") if d == 0 else (a @ b / d).item()


def max_rel_err_pct(dev, gold):
    dev, gold = _t(dev).double(), _t(gold).double()
    return ((dev - gold).abs() / dev.abs().clamp_min(1e-30)).max().item() * 100.0


OPS = ["sum", "mean", "max", "min", "var", "std"]
SCALARS = [2.0, 0.5, -2.0]
N = 32
x = torch.randn(1, 1, N, N, dtype=torch.float32)

print(f"torch {torch.__version__}   input (1,1,{N},{N}) float32, torch.manual_seed(0), dim=-1\n")

for vname, factory in (("A: main @ a4e9bec", _create_golden_function_main),
                       ("B: as rewritten by open PR #56748", _create_golden_function_pr56748)):
    print(f"--- golden variant {vname}")
    hdr = "%-8s %8s %16s %14s   %s" % ("op", "scalar", "max rel err", "PCC", "note")
    print(hdr)
    print("-" * len(hdr))
    for op in OPS:
        g = factory(op)
        for s in SCALARS:
            dev = device(op, x, -1, s)
            gold = g(x, dim=-1, scalar=s)
            note = ""
            if op in ("var", "std"):
                note = "expected %.4gx (scalar^2)" % (s * s) if op == "var" else "expected %.4gx (|scalar|)" % abs(s)
            elif s < 0 and op in ("max", "min"):
                note = "negative scalar flips the extremum: golden returns a DIFFERENT element"
            print("%-8s %8.2f %15.4f%% %14.8f   %s" % (op, s, max_rel_err_pct(dev, gold), pcc(dev, gold), note))
    print()

print("Why this has never failed a test: PCC is invariant under a positive scale factor,")
print("so every PCC-gated comparison still reports a perfect score.")
for op in ("sum", "var"):
    g = _create_golden_function_main(op)
    dev, gold = device(op, x, -1, 2.0), g(x, dim=-1, scalar=2.0)
    print("  %-5s scalar=2.0 -> max rel err %7.2f%%   PCC %.10f" % (op, max_rel_err_pct(dev, gold), pcc(dev, gold)))

# --------------------------------------------------------------- self-check
for factory in (_create_golden_function_main, _create_golden_function_pr56748):
    for op in OPS:
        g = factory(op)
        assert max_rel_err_pct(device(op, x, -1, 2.0), g(x, dim=-1, scalar=2.0)) > 1.0, \
            f"{op}: scalar=2.0 must diverge from the golden"
        # scalar=1.0 is the default and must agree, proving the harness itself is sound
        assert max_rel_err_pct(device(op, x, -1, 1.0), g(x, dim=-1, scalar=1.0)) < 1e-3, \
            f"{op}: scalar=1.0 must agree"
assert pcc(device("sum", x, -1, 2.0), _create_golden_function_main("sum")(x, dim=-1, scalar=2.0)) > 0.99999999, \
    "PCC must stay ~1.0 under a positive scale -> proves PCC cannot catch this"
assert pcc(device("max", x, -1, -2.0), _create_golden_function_main("max")(x, dim=-1, scalar=-2.0)) < 0.5, \
    "negative scalar on max must decorrelate -> proves it is not merely a scale factor"
print("\nself-check: OK")

```

</details>

**2. Input data / link or description**

Generated in-script: `torch.randn(1, 1, 32, 32, dtype=torch.float32)` under `torch.manual_seed(0)`.

**3. Frequency**

100% — deterministic, every call that passes a `scalar` other than `1.0`.

### Environment

**1. Software Versions**

`main` @ `a4e9bec`. torch 2.8.0, Python 3.9.6, macOS arm64. No device needed — the affected code is host-side Python.

**2. Hardware Details**

N/A. The defect is in the host-side golden reference, not a kernel.

### Is this a regression?

No — present since the golden factory was introduced, and still present in the open rewrite PR #56748.

### Impact

`scalar=` silently produces a wrong reference for all six generic reductions in comparison mode, `ttnn.get_golden_function`, and any sweep or fallback that leans on the registered golden. Because the discrepancy is a pure scale factor for positive `scalar`, PCC-gated checks report a perfect 1.0 — so a genuine kernel regression in the scaled path would be masked rather than caught. PR #53854 plans a nightly `scalar=` bit-exactness suite for `sum`; if that lands against this golden it will compare to the unscaled value.

### Suggested fix

One line in the factory (and one field in PR #56748's `_ReductionGoldenSpec`), applied before the dispatch:

```python
def golden_function(input_tensor, dim=None, keepdim=False, correction=None, scalar=1.0, **_):
    if scalar != 1.0:
        input_tensor = input_tensor * scalar
    ...
```

Happy to open the PR (plus a small unit test that pins each of the five contracts above against the registered golden rather than a hand-rolled torch expression) if that is useful.

**Comment by p32929** [comment]
PR up: #57144.

The factory now names `scalar` (default `1.0`) and scales the input before reducing. `+13 / -1`, one file.

Two things I checked before writing it, both of which changed the fix:

1. **Pre-scale is the right semantics, including for negative scalars.** The device post-multiplies for MAX/MIN but flips the op when `scalar < 0` (`generic_reductions.cpp:736`, `:772`) precisely so the result is `reduce(scalar * x)` — which is what `test_max.py:159` (`torch.amax(scalar * x)`) and `test_reduction_min.py:149` assert, with `scalar=-2.5` in the parametrisation. A naive post-multiply in the golden would have been wrong for exactly those cases.
2. **`argmax` shares this factory** but does not bind `scalar` (`argmax_nanobind.cpp:64-66`), so it can never receive one; the `scalar != 1.0` guard makes it an exact no-op there and keeps every default path bit-identical. `topk` has its own factory and is untouched.

Verified device-free against the contracts your own tests pin: all 24 combinations of {sum, mean, max, min, var, std} × {dim -1, -2} × {scalar 2.5, -2.5} now match exactly (`atol=0, rtol=0`); 0 of 24 match today. Worst case is `var` at **84% relative error with PCC = 1.0000000000**, since it scales by `scalar**2` — that is why nothing caught this.

Offer stands on a parametrised test that exercises the *registered golden* rather than a hand-rolled reference, under `tests/ttnn/unit_tests/operations/reduce/` — the absence of one is the actual gap here. Happy to put it in this PR or a follow-up, whichever you prefer.
